### ISSUE TITLE
Integration with Deepmatcher

### ISSUE THREAD

**Issue by rbhatia46** [NilsBarlaug/lemon#1]
Hi,
Thanks for open sourcing the package, interesting idea.

I already have a model trained using deepmatcher package, now I want to introduce another layer of interpretability on top of that, came across your paper, can you please throw some light on how do I use my own custom model built using deepmatcher and interpret new observations using your code ?

Thanks

**Comment by NilsBarlaug** [comment]
Hi Rahul,

Thanks for testing out the package!

To explain the prediction of a pair (or multiple) of your custom matcher you use the `lemon.explain()` function. It is mostly agnostic to the data and the matcher but both has to be adapted to a certain interface.

**The data**
- Must be split into `records_a`, `records_b`, and `record_id_pairs`.
- `records_a` and `records_b` are pandas dataframes with records.
- `record_id_pairs` is a pandas dataframe with two columns `a.rid` and `b.rid` that refer to the indices of `records_a` and `records_b`.

**The matcher**
The `explain()` function simply want a function it can give `records_a`, `records_b`, and `record_id_pairs` to and get back predictions. `records_a`, `records_b`, and `record_id_pairs` follow the same format as the data described above, but will be different from the data you pass (it generates synthetic pairs). The returned predictions can be a pandas series, numpy array, list, etc.

**Deepmatcher**
I don't know your exact setup but I made a working example (worked for the [getting started notebook at the deepmatcher repo](https://nbviewer.jupyter.org/github/anhaidgroup/deepmatcher/blob/master/examples/getting_started.ipynb)). Note: make sure to update to version 0.1.1 of lemon - I fixed a bug to make this example work properly with deepmatcher.

```python
# The code is assuming there already exists a deepmatcher model in a variable named `model`

import pandas as pd
import deepmatcher as dm
import lemon

# Make a wrapper predict function
def predict_proba(records_a, records_b, record_id_pairs):
    pairs = (
        record_id_pairs
        .merge(records_a.add_prefix("left_"), left_on="a.rid", right_index=True)
        .merge(records_b.add_prefix("right_"), left_on="b.rid", right_index=True)
        .drop(columns=["a.rid", "b.rid"])
        .reset_index(drop=True)
        .rename_axis(index="id")
        .reset_index(drop=False)
    )
    # I was able to load the data into a deepmatcher dataset without dumping to csv first
    # Hacky and slow, but it works
    pairs.to_csv("tmp.csv")

    unlabeled = dm.data.process_unlabeled(
      path='tmp.csv',
      trained_model=model
    )
    predictions = model.run_prediction(unlabeled)["match_score"]
    return predictions

# Load some data (from the notebook)
pairs = pd.read_csv('sample_data/itunes-amazon/unlabeled.csv')

# Change it to the lemon format
records_a = pairs[[c for c in pairs.columns if c.startswith("left_")]].rename(columns=lambda c: c[5:]).rename_axis(index="rid")
records_b = pairs[[c for c in pairs.columns if c.startswith("right_")]].rename(columns=lambda c: c[6:]).rename_axis(index="rid")
pair_id_to_explain = 0 #  We'll explain the prediction of the first pair
record_id_pairs = pd.DataFrame({"a.rid": [pair_id_to_explain], "b.rid": [pair_id_to_explain]})

# Generate an explanation
exp = lemon.explain(records_a, records_b, record_id_pairs, predict_proba)

# exp can now be visualized by displaying it in a notebook
# or you can save the explanation to json with exp.save(path)
```
If generating an explanation on your setup is slow try adjusting the `num_samples` argument of the `explain()` function yourself (maybe 100?) - the quality gets lower, but it's faster.

**Interpreting**

The short version is:
- Green: Something the model thinks counts towards a match - i.e., the score would be lower without this present
- Red: Something the model thinks towards a non-match - i.e., the score would be higher without this present
- Gray: Something important the model thinks is missing in the other record - i.e., the match score would increase if something in the other record matched this.

I hope that was helpful? Let me know if you have more questions.

And please let me know how it goes :slightly_smiling_face:

**Comment by rbhatia46** [comment]
@NilsBarlaug Really appreciate the quick response, it was very helpful indeed, will test it out and let you know! Thanks again for the package and providing the boilerplate code for deepmatcher integration👍

**Comment by jprashant21** [comment]
Hi @NilsBarlaug 
Thanks for sharing this package with the community. It is really very useful approach from interpretability point of view. I tried to use the sample code you shared in previous post to use Lemon package with a trained Deepmatcher model and records pair data. But, I am getting **LinAlgError: SVD did not converge in Linear Least Squares** . So, I just wanted to check with you what could be the reason for model not being able to converge and if there is something can be done to resolve the error?

Here is the error traceback:
```
/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py:990: UserWarning: Using non-full backward hooks on a Module that does not return a single Tensor or a tuple of Tensors is deprecated and will be removed in future versions. This hook will be missing some of the grad_output. Please use register_full_backward_hook to get the documented behavior.
  warnings.warn("Using non-full backward hooks on a Module that does not return a "
Finished Epoch 7 || Run Time:    2.8 | Load Time:    1.2 || F1:   0.00 | Prec:   0.00 | Rec:   0.00 || Ex/s:   0.00

Intel MKL ERROR: Parameter 4 was incorrect on entry to DLASCL.

Intel MKL ERROR: Parameter 4 was incorrect on entry to DLASCL.
---------------------------------------------------------------------------
LinAlgError                               Traceback (most recent call last)
/tmp/ipykernel_13867/2219123872.py in <module>
      1 # Generate an explanation
----> 2 exp = lemon.explain(records_a, records_b, record_id_pairs, predict_proba)

/opt/conda/lib/python3.7/site-packages/lemon/_lemon.py in explain(records_a, records_b, record_id_pairs, predict_proba, num_features, dual_explanation, estimate_potential, granularity, num_samples, token_representation, token_patterns, explain_attrs, attribution_method, show_progress, random_state, return_dict)
    636             attribution_method=attribution_method,
    637             show_progress=show_progress and len(record_pairs) == 1,
--> 638             random_state=random_state,
    639         )
    640 

/opt/conda/lib/python3.7/site-packages/lemon/_lemon.py in _explain_record_pair(record_pair, predict_proba, num_features, dual_explanation, estimate_potential, granularity, num_samples, token_representation, token_patterns, explain_attrs, attribution_method, show_progress, random_state)
    553                 random_state=random_state,
    554             )
--> 555             for source in ["a", "b"]
    556         ]
    557         if not math.isclose(explanation_a.prediction_score, explanation_b.prediction_score, rel_tol=1e-2):

/opt/conda/lib/python3.7/site-packages/lemon/_lemon.py in <listcomp>(.0)
    553                 random_state=random_state,
    554             )
--> 555             for source in ["a", "b"]
    556         ]
    557         if not math.isclose(explanation_a.prediction_score, explanation_b.prediction_score, rel_tol=1e-2):

/opt/conda/lib/python3.7/site-packages/lemon/_lemon.py in _explain(record_pair, predict_proba, explain_sources, num_features, num_samples, granularity, token_representation, token_patterns, dual_explanation, estimate_potential, explain_attrs, attribution_method, show_progress, random_state)
    439                 distances,
    440                 num_features,
--> 441                 feature_group_size=(2 if estimate_potential else 1),
    442             )
    443 

/opt/conda/lib/python3.7/site-packages/lemon/_lemon.py in _lime(X, y, distances, num_features, feature_group_size)
    268     y = y - y[0]
    269 
--> 270     used_features = _forward_selection(X, y, sample_weights, num_features, feature_group_size)
    271     X_used = X[:, used_features]
    272     model = LinearRegression(fit_intercept=False)

/opt/conda/lib/python3.7/site-packages/lemon/_lemon.py in _forward_selection(X, y, sample_weights, num_features, feature_group_size)
    249             X_used = X[:, used_features + feature_group]
    250             clf = LinearRegression(fit_intercept=False)
--> 251             clf.fit(X_used, y, sample_weight=sample_weights)
    252             score = clf.score(X_used, y, sample_weight=sample_weights)
    253             if score > max_:

/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/_base.py in fit(self, X, y, sample_weight)
    716                 self._residues = np.vstack([out[3] for out in outs])
    717         else:
--> 718             self.coef_, self._residues, self.rank_, self.singular_ = linalg.lstsq(X, y)
    719             self.coef_ = self.coef_.T
    720 

/opt/conda/lib/python3.7/site-packages/scipy/linalg/basic.py in lstsq(a, b, cond, overwrite_a, overwrite_b, check_finite, lapack_driver)
   1210                                                cond, False, False)
   1211         if info > 0:
-> 1212             raise LinAlgError("SVD did not converge in Linear Least Squares")
   1213         if info < 0:
   1214             raise ValueError('illegal value in %d-th argument of internal %s'

LinAlgError: SVD did not converge in Linear Least Squares
```
## And Here is my code snippet:

```
import pandas as pd
import deepmatcher as dm
import lemon

#### Make a wrapper predict function
def predict_proba(records_a, records_b, record_id_pairs):
    pairs = (
        record_id_pairs
        .merge(records_a.add_prefix("p1_"), left_on="a.rid", right_index=True)
        .merge(records_b.add_prefix("p2_"), left_on="b.rid", right_index=True)
        .drop(columns=["a.rid", "b.rid"])
        .reset_index(drop=True)
        .rename_axis(index="index")
        .reset_index(drop=False)
    )

    pairs.to_csv("data/tmp.csv",index=None)
    ignore_list = ["index", "label"]
    unlabeled = dm.data.process_unlabeled(
      path='data/tmp.csv',
      trained_model=model,
    ignore_columns=ignore_list)
    predictions = model.run_prediction(unlabeled)["match_score"]
    return predictions
    
    
#### Change it to the lemon format
records_a = pairs[[c for c in pairs.columns if c.startswith("p1_")]].rename(columns=lambda c: c[3:]).rename_axis(index="rid")
records_b = pairs[[c for c in pairs.columns if c.startswith("p2_")]].rename(columns=lambda c: c[3:]).rename_axis(index="rid")
pair_id_to_explain = 10 #  We'll explain the prediction of the first pair
record_id_pairs = pd.DataFrame({"a.rid": pair_id_to_explain, "b.rid": pair_id_to_explain})

#### Generate an explanation
exp = lemon.explain(records_a, records_b, record_id_pairs, predict_proba)
```

**Comment by NilsBarlaug** [comment]
Hi @jprashant21,

Thanks for testing out the package!

It is most likely something with the data you're using - for example if the records are empty. It is hard to say what the exact problem is without seeing your data and being able to check it out myself. You said you're using deepmatcher data. Are you able to share what data? If not, please a have a look at the record pair you're trying to explain (`records_a.loc[10]` and `records_b.loc[10]`) and see if there is anything special about it.

Please let me know if you figure it out :slightly_smiling_face:

**Comment by jprashant21** [comment]
Hi @NilsBarlaug 

Thanks for the prompt response. Yes, I'm using deepmatcher model and entity matching data which has almost the same format as your example with `lemon.utils.datasets.deepmatcher.textual_abt_buy()`. I have some pre-processing checks to ensure the data doesn't contain any NaNs or empty strings stuff. I checked (records_a.loc[10] and records_b.loc[10]) as well but doesn't seem to be having anything peculiar. I'm thinking could it be related to `Intel MKL ERROR: Parameter 4 was incorrect on entry to DLASCL.`

**Comment by NilsBarlaug** [comment]
Aha, I see. And I assume this only happens for some record pairs, not all?

I don't really know what the problem is, but I can at least see two potential ways it can happen. Here are some suggestions to fix/debug this from the top of my head:

1. As you suggested, MKL could be the problem or its interaction with scipy/numpy
  - Try using packages without MKL (this should be simple if you're using conda I think)
  - Make sure scipy, numpy, sklearn and MKL are the newest version
2. Maybe the deepmatcher model outputs NaNs or inf?
  - Try printing/checking out `y` in `_forward_selection()` above line 251 (from the stack trace). This is output from your model. Maybe there is NaNs or infs there? In that case, make sure to handle them in some sensible way (set to 0.5?) in the wrapper function before returning it.

I hope you're able to figure it out. Please let me know if there is anything I can do to help you :slightly_smiling_face:

**Comment by jprashant21** [comment]
Hi @NilsBarlaug 
Finally it worked !!!
I had to install and upgrade several packages in my conda env. The issue was related to Intel MKL which was resolved after upgrading mkl, sklearn, scipy, numpy and pandas packages as per your suggestion. Thanks a bunch for your help !!!
